---
title: "Call DNAm haplotypes"
author: "Ks"
date: '`r paste("First created on 10/15/2019. Updated on", Sys.Date())`'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Revised 11/13/2020 to filter out reads with non-converted Cs.

Load Libraries:
```{r libraries, message=FALSE}
library(GenomicAlignments)
library(tidyverse)
library(writexl)
```

# Load annotation

Load target regions, sequences and CpG positions:
```{r readcgpositions}
load("data/cgposWSafterclip.rda")
load("data/chposWSafterclip.rda")
load("data/chposCSafterclip.rda")
load("data/clipgrplus1.rda")
```

Write the clipped target regions to a bed file that I can use for filtering my *.bam files.  The read must cover this entire region for me to read it into memory.
```{r writebed}
bedfl <- cbind.data.frame(seqnames(clipgrplus1),start(clipgrplus1),end(clipgrplus1))
bedfl
write.table(bedfl,file="data/clipgrplus1.bed",sep = "\t",
            row.names = FALSE, col.names = FALSE, quote = FALSE)
```


Filenames for this processing run.
```{r runnames}
#datadir=c("~kims/Google Drive File Stream/My Drive/shibata/methylation bam files 9-19")
#writedatafn=c("~kims/GitHub/runbsseq/data/cmethhaps/darrylsbamfiles9-19")

#datadir=c("~kims/Google Drive File Stream/My Drive/methylation ampliseq bam files 12-19")
#writedatafn=c("~kims/GitHub/newbsseq/data/cmethhaps/experimental-ctls")

datadir=c("~kims/Google Drive File Stream/My Drive/shibata/GalaxyBSbam_6284_6_20")
writedatafn=c("~kims/GitHub/runbsseq/data/cmethhaps/20200620")
```

Read in .bam filenames.
```{r bamfilenames}
bamfiles=dir(datadir,pattern="*\\.bam$")
whichbam <- 55
```

Shorten these filenames for Excel output.
```{r shortenfilenames}
sslist <- strsplit(bamfiles,c("Galaxy"))
snm <- map_chr(sslist, ~.x[1])  
```

Set flags for reading in the data. I will require the mapping quality score is 30 or better.
```{r readGAlign}
#flag0 <- scanBamFlag(isFirstMateRead=TRUE, isSecondMateRead=FALSE,
#                      isDuplicate=FALSE, #isNotPassingQualityControls=FALSE)
#param0 <- ScanBamParam(flag=flag0, what=c("seq","qual"))

flag1 <- scanBamFlag(isFirstMateRead=TRUE, isSecondMateRead=FALSE,
                    isUnmappedQuery=FALSE,isSecondaryAlignment=FALSE,
                    isProperPair=TRUE,isSupplementaryAlignment=FALSE,
                    isNotPassingQualityControls=FALSE)
param1 <- ScanBamParam(which = clipgrplus1,flag=flag1, mapqFilter = 30,
                        what=c("seq","qual"))
```


```{r functions}
# subset gal object to reads covering region
readsonregion <- function(gaobj,region) {
    r1 <- findOverlaps(ranges(gaobj),ranges(region),
                          minoverlap=width(region))
    gal.subset <- gal[queryHits(r1)]
    gal.subset
}

# clip all the reads to the size of my target (1st to last C).
clipreads <- function(reads_on_ref,startpos,region){
  
    clipped.start <- start(region) - startpos + 1
    clippedreads <- BStringSet(reads_on_ref, 
                              start=clipped.start, 
                               width=rep(width(region),
                                         length(reads_on_ref)))
    clippedreads
}

# turn nucleotide vector into pattern of 1/0 with 1 = methylated; 0 = unmethylated
# fnc for the Watson strand (Ws)
biscWs <- function(seqvec,cpxpos){
          hapbase <- unlist(strsplit(toString(seqvec),split=""))[cpxpos]
          hapvec <- ifelse(hapbase=="C",1,ifelse(hapbase=="T",0,NA))
          hapvec
}
# fnc for the Crick strand (Cs)
biscCs <- function(seqvec,cpxpos){
          hapbase <- unlist(strsplit(toString(seqvec),split=""))[cpxpos]
          hapvec <- ifelse(hapbase=="G",1,ifelse(hapbase=="A",0,NA))
          hapvec
}

# turn these into haplotypes 
gethaps <- function(clipgal,thisregion,cpgpos,regiongr,
                    dnamhap){
  nreads <- length(clipgal)
  ncpg <- length(cpgpos)
  haplst <- lapply(clipgal,dnamhap,cpgpos)
  
  hapmat <- data.frame(matrix(NA,nrow=nreads,ncol=ncpg))
  hapmtx <- matrix(unlist(haplst),byrow=T,ncol=ncpg)
  hapmat[,1:ncpg] <- hapmtx
  colnames(hapmat) <- paste("X",1:ncpg,sep="")
   
  datmat <- cbind.data.frame(sample = snm[whichbam],
                             amplicon = thisregion,
                      chr = seqnames(regiongr),
                      start = start(regiongr),
                      readseq = 1:nrow(hapmat),
                      naCpG = rowSums(is.na(hapmat)),
                      hapmat
                      )
  datmat
}
```


```{r bis-non-conversion-functions}
# these are now redundant --test that I can delete them
# call non-converted CpHs in Watson strand
#nccWs <- function(seqvec,cphpos){
#              hapbase <- unlist(strsplit(toString(seqvec),split=""))[cphpos]
#              hapvec <- ifelse(hapbase=="C",1,ifelse(hapbase=="T",0,NA))
#              hapvec
#}
# call non-converted CpHs in Crick strand
#nccCs <- function(seqvec,cphpos) {
#              hapbase <- unlist(strsplit(toString(seqvec),split=""))[cphpos]
#              hapvec <- ifelse(hapbase=="G",1,ifelse(hapbase=="A",0,NA))
#              hapvec
#}

# For each read, call the non-converted CpHs
ncreads <- function(clipgal,cphpos,nccs){
  ncph <- length(cphpos)
  nreads <- length(clipgal)
  
  haplst <- lapply(clipgal,nccs,cphpos)
  haplst
}
```

# DNAm haplotype Pipeline

Now run pipeline for a single sample.  Sorted Bam file + index get saved to local directory. They are deleted after C-methylation haplotype calling. 
```{r sort-index-bam}
start_time <- Sys.time()

sortBam(file.path(datadir,bamfiles)[whichbam],
        paste(snm[whichbam],"sorted",sep="."))
indexBam(paste(snm[whichbam],"sorted.bam",sep="."))
```
 
```{r read-bam} 
gal <- readGAlignments(paste(snm[whichbam],"sorted.bam",sep="."),
              use.names=TRUE, param=param1)
print(paste("# reads input (mapqscore>30 & overlap regions):", length(gal)))
```


```{r count-noncovertedCs, include=FALSE}
#  This chunk is for testing only
#Ws <- which(values(clipgrplus1)$WC_strand=="W")
#For each region do the following: 

#thisregion <- 2
   #Subset to reads that cover single region.
#  gal.sub <- readsonregion(gal,clipgrplus1[thisregion])
#  if(length(gal.sub)>1) {  # require 2 read minimum to report haplotypes
#      qseqs <- mcols(gal.sub)$seq
      # This is to get the sequences aligned to reference with D/Is
#      qseq_on_ref <- sequenceLayer(qseqs, cigar(gal.sub))
#      clipped.qseq <- clipreads(qseq_on_ref,start(gal.sub),
 #                             clipgrplus1[thisregion])

      # I will pass strand information to gethaps so I 
      # know whether to adjust the C(pH) position to the (Cp)H position. 
      # If C strand, then use H position; W strand uses C position.
#      i <- match(thisregion,Ws,nomatch=0)
#      if (i) { nccslist <- ncreads(clipped.qseq,
 #                         chposWSafterclip[[i]],
#                          nccWs)
#      }  else {
#          i <- match(thisregion,Cs)
#          nccslist <- ncreads(clipped.qseq,
#                          chposCSafterclip[[i]] ,
#                          nccCs)
#      }
      # summarize output then add filter later.
#       ncsum  <- cbind.data.frame(numnc = unlist(lapply(nccslist,sum)), ncph = unlist(lapply(nccslist,length)) )
#       print(paste("Region",thisregion,":"))
#       print(table(ncsum$numnc,useNA="ifany"))
       
#       print(sum(table(ncsum$numnc,useNA="ifany")))
#       print(length(clipped.qseq))
```

Analysis in 3 steps:  
Step 1. Call the Cs  as 1/0 (1 = not-converted; 0 = converted)
Step 2. Summarize Quality of bisulfite conversion/C sequencing errors
Step 3. Filter and create haplotype patterns  
```{r call-Cs}
# create empty list of data frames for each amplicon 
my.names<-paste("amplicon",1:38,sep="")
ldf <- setNames(replicate(38,data.frame()),my.names)

# amplicon is on which strand? Watson or Crick
Ws <- which(values(clipgrplus1)$WC_strand=="W")
Cs <- which(values(clipgrplus1)$WC_strand=="C")

#For each region do the following: 
for (thisregion in 1:38) {
   #Subset to reads that cover single region.
  gal.sub <- readsonregion(gal,clipgrplus1[thisregion])
    if(length(gal.sub)>1) {  # require 2 read minimum to report haplotypes
      qseqs <- mcols(gal.sub)$seq
      # This is to get the sequences aligned to reference with D/Is
      qseq_on_ref <- sequenceLayer(qseqs, cigar(gal.sub))
      clipped.qseq <- clipreads(qseq_on_ref,start(gal.sub),
                              clipgrplus1[thisregion])

      # I first need to know what strand the amplicon is on to 
      # know whether to adjust the C(pG) position to the (Cp)G position. 
      # If C strand, then use G position; W strand uses C position.
      
      i <- match(thisregion,Ws,nomatch=0)
      if (i) { #first call non-converted Cs as 1=nonconverted; 0=converted
              nccslist <- ncreads(clipped.qseq,
                          chposWSafterclip[[i]],
                          biscWs)
               #then record CpG methylation as 1/0
               dmat <- gethaps(clipped.qseq,
                      thisregion,
                      cgposWSafterclip[[thisregion]],
                      clipgrplus1[thisregion],
                      biscWs)
      }  else {
          i <- match(thisregion,Cs)
          nccslist <- ncreads(clipped.qseq,
                          chposCSafterclip[[i]],
                          biscCs)
          dmat <- gethaps(clipped.qseq,
                      thisregion,
                      as.integer(cgposWSafterclip[[thisregion]]+1),
                      clipgrplus1[thisregion],
                      biscCs)
      }
      #  count the number of non-converted Cs in each read
       ncsum  <- cbind.data.frame(numnc = unlist(lapply(nccslist,sum)), 
                                  ncph = unlist(lapply(nccslist,length)) )
       
      ldf[[thisregion]] <- cbind.data.frame(ncsum,dmat)    
    }
}

```

Now report on QC of the individual reads by Amplicon:

```{r summarize-quality-then-filter}
numberreadsonregion <- sum(map_int(ldf,nrow))
print(paste("Total number of Reads:",numberreadsonregion))

qcmat <- matrix(rep(NA,4*38),ncol=4)
rownames(qcmat) <- paste0("amplicon",1:38)
colnames(qcmat) <- c("NumReads","NumNCReads","NumSeqErCpGReads","FractFilt")
for (i in  1:38) {
    dmat <- ldf[[i]]
    print(paste("Region",i,":"))
    print("Non-converted Cs:    (NA is sequencing error)")
    print(table(dmat$numnc,useNA="ifany"))
    ncreads <- ifelse(is.na(dmat$numnc) | (dmat$numnc > 0),1,0)
    ncpgseqerr <- ifelse( dmat$naCpG  > 0, 1, 0)
    qcmat[i,"NumReads"] <- nrow(dmat)
    qcmat[i,"NumNCReads"] <- sum(ncreads)
    qcmat[i,"NumSeqErCpGReads"] <- sum(ncpgseqerr)
    frfilt <- sum(ifelse((ncreads + ncpgseqerr) > 0,1,0))/qcmat[i,"NumReads"]
    qcmat[i,"FractFilt"] <- round(frfilt,3)
}
print(qcmat)
```

Step 3: Remove reads with non-converted Cs or sequencing errors, summarize once more, and write to datafile.

```{r filter-reads}

for (i in  1:38) { 
  dmat <- ldf[[i]]
  
  # filter on non-converted reads and reads with sequencing error (missing data) at CpGs 
  if (nrow(dmat)>1) {   # need more than 1 read to measure hamming distance
      dmat <- dmat %>%
            filter(!is.na(numnc) & numnc == 0 & naCpG==0)
  
      if (nrow(dmat)>1) { # need more than 1 read to measure hamming distance
          # Remove the first 8 columns and create htype
          hapmat <- dmat[,-c(1:8)]
          ncpg <- ncol(hapmat)
  
          if (ncpg>1) {
             htype <- apply(hapmat[,(1:ncpg)],1,paste,collapse="")
          } else {
             htype <- as.character(hapmat[,1])
          }
          datmat <- cbind.data.frame(dmat[,c(1:8)],
                                      htype = htype,
                                      hapmat)
          ldf[[i]] <- datmat
      }
  }  
}

end_time <- Sys.time()
end_time - start_time
```

is the run time for sample 
```{r samplename}
snm[whichbam]
```

# Read Depths/Region

Here's the read depth for each amplicon.
```{r amplicondepth}
map_int(ldf,nrow)
```

And total read depth:
```{r totreaddepth}
sum(map_int(ldf,nrow))
```

And the fraction of mapped reads we retained:
```{r fractionreadsretained}
sum(map_int(ldf,nrow))/numberreadsonregion
```

```{r writefiles}
snm2 <- paste(snm,"xlsx",sep=".")
write_xlsx(ldf,file.path(writedatafn,snm2[whichbam]) )

file.remove(paste(snm[whichbam],"sorted.bam",sep="."))
file.remove(paste(snm[whichbam],"sorted.bam.bai",sep="."))
```

# Excel File created:
```{r whichfile}
snm2[whichbam]
```

```{r sessionInfo}
sessionInfo()
```